src/main/java/com/arpnetworking/tsdcore/sinks/HttpSinkActor.java

```diff
@@ -306,6 +323,31 @@ private void fireNextRequest() {
         PatternsCS.pipe(responsePromise, context().dispatcher()).to(self());
     }
 
+    private HttpResponse sendHttpRequest(
+            final Request request,
+            final int attempt) {
+        try {
+            Thread.sleep(Double.valueOf(Math.pow(2, attempt) - 1).longValue() * 1000);
```

Comment:
You can't do a thread sleep here. It can block too much. The rule is basically not to block in an asyc method.

Also, this isn't proper exponential backoff. Please take a look at https://en.m.wikipedia.org/wiki/Exponential_backoff

Specifically the part about randomization.

Reply:
By saying "block too much", do you mean 1) it can block more than the time set or 2)it occupies too much resources because it is a busy wait? 
I found a solution [here](https://stackoverflow.com/questions/24104313/how-do-i-make-a-delay-in-java) using `ScheduledExecutorService` which can solve 1). 
And we can probably use wait() for a non busy wait if the problem is 2).

Reply:
You need to understand how this code is executing. This code is in an Akka actor, which means it is running on a thread pool shared with other actors. There is a limited amount of threads. In an actor, you should not do anything that blocks the thread (in most cases, there can be exceptions if you are very careful and configure things explicitly). But this also gives us access to Akka timers where you can schedule execution in a non blocking way. Please take a look at the following documentation. 

https://doc.akka.io/docs/akka/current/actors.html#introduction

https://doc.akka.io/docs/akka/current/dispatchers.html

https://doc.akka.io/docs/akka/current/actors.html#actors-timers